### PR TITLE
Make TLS accept logic compatible with disabled protocol detection

### PR DESCRIPTION
**Pull request by hawkw**
This branch changes the proxy's accept logic so that the proxy will no
longer attempt to terminate TLS on ports which are configured to skip
protocol detection. This means that a Linkerd deployment with 
`--tls optional` will no longer break server-speaks-first protocols like
 MySQL (although that traffic will not be encrypted). 

Since it's necessary to get the connection's original destination to
determine if it's on a port which should skip protocol detection, I've
moved the SO_ORIGINAL_DST call down the stack from `Server` to
`BoundPort`. However, to avoid making an additional unnecessary syscall,
the original destination is propagated to the server, along with the
information about whether or not protocol detection is enabled. This is
the approach described in
https://github.com/linkerd/linkerd2/issues/1270#issuecomment-406124236.

I've also written a new integration test for server-speaks-first
protocols with TLS enabled. This test is essentially the same as the
existing `transparency::tcp_server_first` test, but with TLS enabled for
the test proxy. I've confirmed that this fails against master.
Furthermore, I've validated this change by deploying the `booksapp` demo
with MySQL with TLS enabled, which [previously didn't work](https://github.com/linkerd/linkerd2/issues/1648#issuecomment-432867702).

Closes linkerd/linkerd2#1270

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

### CODE REVIEW

**Comment by hawkw** [comment]
@klingerf if you're interested in validating this change, it fixes the issues you described in https://github.com/linkerd/linkerd2/issues/1648#issuecomment-432867702 :)

**Comment by hawkw** [comment]
As a side note, this change basically means that all the SO_ORIGINAL_DST code is encapsulated inside the `transport` module.

File: src/proxy/server.rs [hunk 1]
```diff
@@ -251,21 +242,13 @@ where
             _p: (),
         };
 
+
```
Comment (dadjeibaah) [inline src/proxy/server.rs]: Probably the first valuable review comment I can make with this rust code. Extra whitespace 😂

File: src/transport/connection.rs [hunk 2]
```diff
@@ -14,11 +14,14 @@ use tokio::{
 
 use Conditional;
 use transport::{AddrInfo, BoxedIo, GetOriginalDst, tls};
+use indexmap::IndexSet;
 
-pub struct BoundPort {
-    inner: std::net::TcpListener,
+pub struct BoundPort<G = ()> {
+    inner: Option<std::net::TcpListener>,
```
Comment (dadjeibaah) [inline src/transport/connection.rs]: Out of curiousity, why does this need to be wrapped inside an option?

Comment (hawkw) [reply src/transport/connection.rs]: Yeah, so that's because we need to [move](https://doc.rust-lang.org/book/ch04-01-what-is-ownership.html#ways-variables-and-data-interact-move) the socket into [`tokio::net::TcpListener::from_std`](https://tokio-rs.github.io/tokio/tokio/net/tcp/struct.TcpListener.html#method.from_std) (which takes a `std::net::TcpListener` _by value_) here:
https://github.com/linkerd/linkerd2-proxy/blob/8a9e01ec503dbeb938e51860c3562c592c08a72e/src/transport/connection.rs#L230
but, we now call the new `new_conn` method on a reference to the struct (`&self`) in the closure passed to `Stream::and_then` here:
https://github.com/linkerd/linkerd2-proxy/blob/8a9e01ec503dbeb938e51860c3562c592c08a72e/src/transport/connection.rs#L239-L248

Since the listener can no longer be moved out of the struct by deconstructing it, as the struct still needs to be alive when the closure is invoked, we now use [`Option::take`](https://doc.rust-lang.org/std/option/enum.Option.html#method.take), which moves the value out of the `Option`, replacing it with `None`. We can use this to move the `TcpListener` out of the `BoundPort` struct without dropping the `BoundPort`:
https://github.com/linkerd/linkerd2-proxy/blob/8a9e01ec503dbeb938e51860c3562c592c08a72e/src/transport/connection.rs#L222-L223
since `Option::take` takes the receiver by a _mutable reference_ (`&mut self`). 

This is a fairly common pattern in Rust when dealing with similar situations.

File: src/transport/connection.rs [hunk 3]
```diff
@@ -412,6 +518,10 @@ impl Connection {
     pub fn tls_status(&self) -> tls::Status {
         self.tls_status
     }
+
+    pub fn disable_protocol_detection(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (hawkw) [reply src/transport/connection.rs]: I had originally called this `should_disable_protocol_detection` but that seemed a little too wordy; will rename it.

**Comment by hawkw** [comment]
I'm trying to figure out why the CI build for 38782e30327d010a5e9f61accf9e504abd09d0f7 is failing. It seems unrelated to the change (and builds on my machine):

```
error: overflow representing the type `*const alloc::sync::ArcInner<support::linkerd2_proxy::proxy::http::glue::HyperConnect<support::linkerd2_proxy::transport::metrics::Connect<linkerd2_timeout::Timeout<support::linkerd2_proxy::transport::connect::Target>>>>`
```

The really curious thing is that 38782e30327d010a5e9f61accf9e504abd09d0f7 only changes formatting and names, and the previous commit (with functionally equivalent code) passed CI...

**Comment by hawkw** [comment]
It seems like the CI error was caused by the names of certain fields on `Connection` causing some internal storage in the compiler to overflow when a `Connection` is deeply nested in another type. What's interesting about this is that apparently the size of that storage depends on the environment where the code is being compiled, because the crash occurs on CI but I can't trigger it on my machine. 

I've fixed the issue by shortening some of the field names on `Connection`. @seanmonstar, since you previously had feedback on naming, is the new naming (as of ed31188dcf1023ed994a55423c17284f60755669) still acceptable?
